File: RegEx.rb
class RegEx
    def isNullable() raise "" end
    def derive(c) raise "" end
    def matches(string)
        if string.size == 0
            self.isNullable
        else
            self.derive(string[0]).matches(string[1..-1])
        end
    end
end

class Empty < RegEx
    def isNullable
        false
    end
    def derive(c)
        self
    end
end

class Blank < RegEx
    def isNullable
        true
    end
    def derive(c)
        Empty.new
    end
end

class Primitive < RegEx
    def initialize(c)
        @c = c
    end
    def isNullable
        false
    end
    def derive(c)
        if @c == c
            Blank.new
        else
            Empty.new
        end
    end
end

class Choice < RegEx
    def initialize(this, that)
        @this = this
        @that = that
    end
    def isNullable
        @this.isNullable || @that.isNullable
    end
    def derive(c)
        Choice.new(@this.derive(c), @that.derive(c))
    end
end

class Repetition < RegEx
    def initialize(base)
        @base = base
    end
    def isNullable
        true
    end
    def derive(c)
        Sequence.new(@base.derive(c), self)
    end
end

class Sequence < RegEx
    def initialize(left, right)
        @left = left
        @right = right
    end
    def isNullable
        @left.isNullable && @right.isNullable
    end
    def derive(c)
        if @left.isNullable
            Choice.new(Sequence.new(@left.derive(c), @right), @right.derive(c))
        else
            Sequence.new(@left.derive(c), @right)
        end
    end
end

class Intersection < RegEx
    def initialize(this, that)
        @this = this
        @that = that
    end
    def isNullable
        @this.isNullable && @that.isNullable
    end
    def derive(c)
        Intersection.new(@this.derive(c), @that.derive(c))
    end
end

class Difference < RegEx
    def initialize(left, right)
        @left = left
        @right = right
    end
    def isNullable
        @left.isNullable && !@right.isNullable
    end
    def derive(c)
        Difference.new(@left.derive(c), @right.derive(c))
    end
end

class Complement < RegEx
    def initialize(base)
        @base = base
    end
    def isNullable
        !@base.isNullable
    end
    def derive(c)
        Complement.new(@base.derive(c))
    end
end

puts Empty.new.matches("anything")
puts Blank.new.matches("anything")
puts Blank.new.matches("")
puts Primitive.new('c').matches('c')
f = Primitive.new 'f'
puts f.matches 'f'
o = Primitive.new 'o'
foo = Sequence.new(f, Sequence.new(o, o))
puts foo.matches("foo")
puts foo.matches("bar")
b = Primitive.new 'b'
a = Primitive.new 'a'
r = Primitive.new 'r'
bar = Sequence.new(b, Sequence.new(a, r))
puts bar.matches 'bar'
foobar = Choice.new(foo, bar)
puts foobar.matches 'foo'
puts foobar.matches 'bar'
puts foobar.matches 'foobar'

foobarstar = Repetition.new foobar
puts foobarstar.matches 'foobarfoo'
puts foobarstar.matches 'foobarfoof'

File: RegEx.py
class RegEx:
    def isNullable(self): raise Exception()
    def derive(self, c): raise Exception()
    def matches(self, string):
        if len(string) == 0:
            return self.isNullable()
        else:
            return self.derive(string[0]).matches(string[1:])

class Empty(RegEx):

    def isNullable(self):
        return False

    def derive(self, c):
        return self

class Blank(RegEx):

    def isNullable(self):
        return True

    def derive(self, c):
        return Empty()

class Primitive(RegEx):

    def __init__(self, c):
        self.c = c

    def isNullable(self):
        return False

    def derive(self, c):
        if self.c == c:
            return Blank()
        else:
            return Empty()

class Choice(RegEx):
    def __init__(self, this, that):
        self.this = this
        self.that = that

    def isNullable(self):
        return self.this.isNullable() or self.that.isNullable()

    def derive(self, c):
        return Choice(self.this.derive(c), self.that.derive(c))

class Repetition(RegEx):
    def __init__(self, base):
        self.base = base

    def isNullable(self):
        return True

    def derive(self, c):
        return Sequence(self.base.derive(c), self)

class Sequence(RegEx):

    def __init__(self, left, right):
        self.left = left
        self.right = right

    def isNullable(self):
        return self.left.isNullable() and self.right.isNullable()

    def derive(self, c):
        if (self.left.isNullable()):
            return Choice(Sequence(self.left.derive(c), self.right),
                          self.right.derive(c))
        else:
            return Sequence(self.left.derive(c), self.right)

class Intersection(RegEx):

    def __init__(self, this, that):
        self.this = this
        self.that = that

    def isNullable(self):
        return self.this.isNullable() and self.that.isNullable()

    def derive(self, c):
        return Intersection(self.this.derive(c), self.that.derive(c))

class Difference(RegEx):

    def __init__(self, left, right):
        self.left = left
        self.right = right

    def isNullable(self):
        return self.left.isNullable() and not self.right.isNullable()

    def derive(self, c):
        return Difference(self.left.derive(c), self.right.derive(c))

class Complement(RegEx):

    def __init__(self, base):
        self.base = base

    def isNullable(self):
        return not base.isNullable()

    def derive(self, c):
        return Complement(self.base.derive(c))

# tests
print(Empty().matches("anything"))
print(Blank().matches("anything"))
print(Blank().matches(""))
print(Primitive('c').matches("c"))
f = Primitive('f')
print(f.matches('f'))
o = Primitive('o')
foo = Sequence(f, Sequence(o,o))
print(foo.matches("foo"))
print(foo.matches("bar"))
b = Primitive('b')
a = Primitive('a')
r = Primitive('r')
bar = Sequence(b, Sequence(a,r))
print(bar.matches("bar"))

foobar = Choice(foo, bar)
print(foobar.matches("foo"))
print(foobar.matches("bar"))
print(foobar.matches("foobar"))

foobarstar = Repetition(foobar)
print(foobarstar.matches("foobarfoo"))
print(foobarstar.matches("foobarfoof"))
